src/ace-adapter.ts

```diff
@@ -0,0 +1,318 @@
+    }
+    let restLength = this.lastDocLines.join("\n").length - start;
+    if (change.action === "remove") {
+      restLength -= text.length;
+    }
+    const insert_op = new TextOperation()
+      .retain(start, null)
+      .insert(text, null)
+      .retain(restLength, null);
+    const delete_op = new TextOperation()
+      .retain(start, null)
+      .delete(text)
+      .retain(restLength, null);
+    if (change.action === "remove") {
+      return [delete_op, insert_op];
+    } else {
+      return [insert_op, delete_op];
+    }
+  }
+
+  private applyOperationToACE(operation: ITextOperation): void {
+    let index = 0;
+    for (const op of operation.getOps()) {
+      if (op.isRetain()) {
+        index += op.chars ?? 0;
+      } else if (op.isInsert()) {
+        this.aceDoc.insert(this.posFromIndex(index), op.text ?? "");
+        index += op.text?.length ?? 0;
+      } else if (op.isDelete()) {
+        const from = this.posFromIndex(index);
+        const to = this.posFromIndex(index + (op.chars ?? 0));
+        const range = this.aceRange.fromPoints(from, to);
+        this.aceDoc.remove(range);
+      }
+    }
+    this.grabDocumentState();
+  }
+
+  private posFromIndex(index: number): Ace.Position {
+    for (const [row, line] of this.aceDoc["$lines"].entries()) {
+      if (index <= line.length) {
+        return { row: row, column: index };
+      }
+      index -= line.length + 1;
+    }
+    return { row: 0, column: index };
+  }
+
+  private indexFromPos(pos: Ace.Position, lines?: string[]): number {
+    lines = lines || this.lastDocLines;
+    let index = 0;
+    for (let i = 0; i < pos.row; i++) {
+      index += this.lastDocLines[i].length + 1;
+    }
+    index += pos.column;
+    return index;
+  }
+
+  private addStyleRule(css: string): void {
+    if (!document) {
+      return;
+    }
+    if (!this.addedStyleRules) {
+      this.addedStyleRules = {};
+      const styleElement = document.createElement("style");
+      document.documentElement
+        .getElementsByTagName("head")[0]
+        .appendChild(styleElement);
+      this.addedStyleSheet = styleElement.sheet;
+    }
+    if (this.addedStyleRules[css]) {
+      return;
+    }
+    this.addedStyleRules[css] = true;
+    this.addedStyleSheet?.insertRule(css, 0);
+  }
+
+  private trigger(event: string, ...args: TextOperation[]): void {
+    if (this.callbacks && this.callbacks[event]) {
+      this.callbacks[event].apply(this, args);
+    }
+  }
+
+  applyOperation(operation: ITextOperation): void {
+    this.ignoreChanges = !operation.isNoop();
+    this.applyOperationToACE(operation);
+    this.ignoreChanges = false;
+  }
+
+  registerUndo(undoFn: UndoRedoCallbackType): void {
+    this.ace.undo = undoFn;
+  }
+
+  registerRedo(redoFn: UndoRedoCallbackType): void {
+    this.ace.redo = redoFn;
+  }
+
+  invertOperation(operation: ITextOperation): ITextOperation {
+    return operation.invert(this.getValue());
+  }
+
+  getText(): string {
```

Comment:
can remove monaco specific methods from `IEditorAdapter`. Can create a separate IEditorAdapter for monaco (implementing base `IEditorAdapter`).

Reply:
Skipping for now as discussed. Also implemented setText and getText methods to set default text here